File: Numerical_Methods/Numerical Integration Techniques/Weddle’s Rule/java_solutions/Problem7.java
public class Problem7 {
    public static double f(double x) {
        return x * x * x + 2 * x * x + 3 * x + 4;
    }

    public static double weddlesRule(double a, double b, int n) {
        if (n % 6 != 0) {
            throw new IllegalArgumentException("Number of intervals must be divisible by 6");
        }

        double h = (b - a) / n;
        double sum1 = f(a) + f(b);  // First and last terms
        double sum2 = 0, sum3 = 0, sum4 = 0, sum5 = 0;

        for (int i = 1; i < n; i++) {
            double x = a + i * h;
            if (i % 6 == 1) {
                sum2 += f(x);
            } else if (i % 6 == 2 || i % 6 == 4) {
                sum3 += f(x);
            } else if (i % 6 == 3) {
                sum4 += f(x);
            } else if (i % 6 == 5) {
                sum5 += f(x);
            }
        }

        sum2 *= 5;
        sum4 *= 6;
        sum5 *= 5;

        return (3 * h / 10) * (sum1 + sum2 + sum3 + sum4 + sum5);
    }

    public static void main(String[] args) {
        double a = -2;  // Lower limit
        double b = 2;   // Upper limit
        int n = 24;     // Number of intervals

        try {
            double result = weddlesRule(a, b, n);
            double exactValue = (Math.pow(b, 4)/4 + 2*Math.pow(b, 3)/3 + 3*b*b/2 + 4*b) - 
                              (Math.pow(a, 4)/4 + 2*Math.pow(a, 3)/3 + 3*a*a/2 + 4*a);
            double error = Math.abs(result - exactValue);

            System.out.printf("Approximate value using Weddle's Rule: %.6f%n", result);
            System.out.printf("Exact value: %.6f%n", exactValue);
            System.out.printf("Absolute error: %.6f%n", error);
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
} 

File: Numerical_Methods/Numerical Integration Techniques/Weddle’s Rule/python_solutions/problem7.py
import numpy as np

def f(x):
    return x**3 + 2*x**2 + 3*x + 4

def weddles_rule(f, a, b, n):
    if n % 6 != 0:
        raise ValueError("Number of intervals must be divisible by 6")
    
    h = (b - a) / n
    x = np.linspace(a, b, n+1)
    y = f(x)
    
    # Weddle's Rule formula
    sum1 = y[0] + y[-1]  # First and last terms
    sum2 = 5 * np.sum(y[1:-1:6])  # Terms at positions 1,7,13,...
    sum3 = y[2:-1:6] + y[4:-1:6]  # Terms at positions 2,8,14,... and 4,10,16,...
    sum4 = 6 * np.sum(y[3:-1:6])  # Terms at positions 3,9,15,...
    sum5 = 5 * np.sum(y[5:-1:6])  # Terms at positions 5,11,17,...
    
    integral = (3*h/10) * (sum1 + sum2 + np.sum(sum3) + sum4 + sum5)
    return integral

def main():
    a = -2  # Lower limit
    b = 2   # Upper limit
    n = 24  # Number of intervals
    
    result = weddles_rule(f, a, b, n)
    exact_value = (b**4/4 + 2*b**3/3 + 3*b**2/2 + 4*b) - (a**4/4 + 2*a**3/3 + 3*a**2/2 + 4*a)
    error = abs(result - exact_value)
    
    print(f"Approximate value using Weddle's Rule: {result:.6f}")
    print(f"Exact value: {exact_value:.6f}")
    print(f"Absolute error: {error:.6f}")

if __name__ == "__main__":
    main() 